go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+// corresponding DNS name along with the associated addresses.
+func (extEgDNS *ExternalEgressDNS) reconcileDNSNameResolver(key string) error {
+	// Split the key in namespace and name of the corresponding object.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("reconcileDNSNameResolver failed to split meta namespace cache key %s for dns name resolver: %v", key, err)
+		return nil
+	}
+	// Fetch the dns name resolver object using the name and namespace.
+	obj, err := extEgDNS.dnsLister.DNSNameResolvers(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// DNSNameResolver object was deleted.
+
+			// Get the corresponding DNS name, if avialable.
+			dnsName, found := extEgDNS.resolverToDNSName[name]
+			// If the DNS name was not found then return.
+			if !found {
+				return nil
+			}
+
+			// Delete the address set associated with the DNS name.
+			var err error
+			waitErr := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (bool, error) {
+				deleteResp := extEgDNS.Delete(DeleteRequest{DNSName: dnsName})
+				err = deleteResp.Err
+				if err != nil {
+					return false, nil
+				}
+				return true, nil
+			})
+			if waitErr != nil {
+				errs := []error{waitErr}
+				if err != nil {
+					errs = append(errs, err)
+				}
+				return fmt.Errorf("error with reconcileDNSNameResolver - %v", errors.NewAggregate(errs))
+			}
+
+			// Delete the corresponding entries from the objDNSName
+			// and the dnsNameObj maps.
+			delete(extEgDNS.resolverToDNSName, name)
+			delete(extEgDNS.dnsNameToResolver, dnsName)
+
+			return nil
+		}
+		return fmt.Errorf("reconcileDNSNameResolver failed to fetch dns name resolver %s in namespace %s", name, namespace)
+	}
+
+	// DNSNameResolver object was added/updated.
+
+	// Check if the object name corresponding to the DNS name exists
+	// or not. If it exists, whether the object name matches with the
+	// exisiting extry.
+	dnsName := string(obj.Spec.Name)
+	objName, exists := extEgDNS.dnsNameToResolver[dnsName]
+	if exists && objName != name {
+		return nil
+	}
+
+	// Add the detials of the object name and the DNS name to the
+	// objDNSName and the dnsNameObj maps.
+	extEgDNS.resolverToDNSName[name] = dnsName
+	extEgDNS.dnsNameToResolver[dnsName] = name
+
+	// Get the addresses corresponding to the DNS name and add them to
+	// the address set corresponding to the DNS name.
+	addresses := []string{}
+	for _, resolvedName := range obj.Status.ResolvedNames {
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			addresses = append(addresses, resolvedAddress.IP)
+		}
+	}
+	addResp := extEgDNS.Add(AddRequest{DNSName: dnsName, Addresses: addresses})
+	return addResp.Err
+
+}
+
+// Add is called whenever a DNS name is needed to be added or updated.
+func (extEgDNS *ExternalEgressDNS) Add(request AddRequest) AddResponse {
```

Comment:
we can 2 things on add here:
1. create and update address set on DNSResolver events, wait here when the address set will be created
2. use `EnsureAddressSet` to only create address set if it doesn't exist, and trust a DNSResolver event to fill the IPs later
I don't think we should return error on `AddNamespace` if the name doesn't exist yet, because it is expected to happen sometimes when cluster manager is slower than node, so I would prefer one the ^ given options

Reply:
Moved the code to create the resolved name along with the creation of address set to `ensureResolvedName()`. This function is called from both `addDNSName()` and `Add()` functions. Thus, the error is returned only during the creation of the address set.